### PR TITLE
added summary and partial validation info

### LINKED ISSUE #933: Notify users about partial validations
# Overview

Currently, there is a programmatic flag `report.tasks[n].partial == True` indicating that the validation task stopped before the end of the file hitting one of the limits error/row/memory/etc

It will be better to also notify end-users about it e.g. with a `warning`.

### LINKED ISSUE #1094: Add validation summary to the CLI report
# Overview

Currently, we have:

```
# -------
# invalid: data/invalid.csv
# -------

===  =====  ===============  ====================================================================================
row  field  code             message
===  =====  ===============  ====================================================================================
         3  blank-label      Label in the header in field at position "3" is blank
         4  duplicate-label  Label "name" in the header at position "4" is duplicated to a label: at position "2"
  2      3  missing-cell     Row at position "2" has a missing cell in field "field3" at position "3"
  2      4  missing-cell     Row at position "2" has a missing cell in field "name2" at position "4"
  3      3  missing-cell     Row at position "3" has a missing cell in field "field3" at position "3"
  3      4  missing-cell     Row at position "3" has a missing cell in field "name2" at position "4"
  4         blank-row        Row at position "4" is completely blank
  5      5  extra-cell       Row at position "5" has an extra value in field at position "5"
===  =====  ===============  ====================================================================================
```

We'd like to extend the report to something like:

```
# -------
# invalid: data/invalid.csv
# -------

# Summary

[table with stats]

# Errors

===  =====  ===============  ====================================================================================
row  field  code             message
===  =====  ===============  ====================================================================================
         3  blank-label      Label in the header in field at position "3" is blank
         4  duplicate-label  Label "name" in the header at position "4" is duplicated to a label: at position "2"
  2      3  missing-cell     Row at position "2" has a missing cell in field "field3" at position "3"
  2      4  missing-cell     Row at position "2" has a missing cell in field "name2" at position "4"
  3      3  missing-cell     Row at position "3" has a missing cell in field "field3" at position "3"
  3      4  missing-cell     Row at position "3" has a missing cell in field "name2" at position "4"
  4         blank-row        Row at position "4" is completely blank
  5      5  extra-cell       Row at position "5" has an extra value in field at position "5"
===  =====  ===============  ====================================================================================
```

# Plan

- [ ] brainstorm the summary table contents
- [ ] agree with @roll @shashigharti 
- [ ] implement

### PR DESCRIPTION
**Pull request by shashigharti**
* Added summary file and errors
* Added "partial info" in validation result to indicate partial validation
* Added new tests
* Fixed existing test those were failing

- fixes #1094
- fixes #933

---
Feedback required:
Overall flow

### CODE REVIEW

**Comment by roll** [comment]
Hi @shashigharti, can you please add a screenshot/output?

**Comment by shashigharti** [comment]
Hi @roll 
Here are the screenshots:
Summary
summary

Summary with partial info
partial

**Comment by shashigharti** [comment]
Screenshot for valid resource:
valid

**Comment by roll** [comment]
@shashigharti 
Thanks! I've review on Monday.

One thing regarding "partial" - I really doubt that anyone except us will be able to understand what we do mean here. I would think about a sentence that clearly explains the situation like "The validation has been done partially because of one of the limits / or similar"

**Comment by shashigharti** [comment]
Thanks! I have made the changes. 
errors

File: frictionless/helpers.py [hunk 1]
```diff
@@ -631,3 +632,41 @@ def dicts_to_markdown_table(dicts: List[dict], **kwargs) -> str:
         error = errors.GeneralError(note="Please install `pandas` package")
         raise module.FrictionlessException(error)
     return df.where(df.notnull(), None).to_markdown(index=False)
+
+
+def validation_summary(
```
Comment (roll) [inline frictionless/helpers.py]: I think it's better to have in the `program.validate` module to be close to the place where it's used. We can use a simple logic for functions like this -- in how many places this function is used? if in only one place, it's a good change we don't need to move it into helpers as helpers is a reusable generic library

File: frictionless/program/validate.py [hunk 2]
```diff
@@ -263,7 +262,42 @@ def program_validate(
...
+            rows_checked=rows_checked,
+            error_list=error_list,
+        )
+        typer.echo("\n# Summary \n")
+        if task.partial:
+            typer.echo(
+                "The document was partially validated because of one of the limits"
+            )
+            typer.echo("* limit errors")
+            typer.echo("* memomry Limit \n")
```
Comment (roll) [inline frictionless/program/validate.py:286]: memomry

File: frictionless/helpers.py [hunk 3]
```diff
@@ -631,3 +632,41 @@ def dicts_to_markdown_table(dicts: List[dict], **kwargs) -> str:
...
+        content.append(["Rows Checked(Partial)**", rows_checked])
+    if error_list:
+        content.append(["Total Errors", sum(error_list.values())])
+    for code, count in error_list.items():
+        content.append([code, count])
+
+    return content
+
+
+def format_bytes(size: int):
```
Comment (roll) [inline frictionless/helpers.py]: E.g. this function is clearly a generic one so I think putting it to `helpers` is a good idea

File: frictionless/program/validate.py [hunk 4]
```diff
@@ -106,6 +105,9 @@ def program_validate(
     pick_errors = helpers.parse_csv_string(pick_errors)
     skip_errors = helpers.parse_csv_string(skip_errors)
 
+    # Summary
+    timer = helpers.Timer()
```
Comment (roll) [inline frictionless/program/validate.py]: Can we use the time data from the report?

**Review by roll — CHANGES_REQUESTED** [review]
Thanks!

It looks good, just a few minor comments.

In general, I think we will continue tidying up the CLI messages for v5 so it's a good step forward

File: frictionless/helpers.py (hunk 1 above)
Comment (shashigharti) [reply frictionless/helpers.py]: Sure! I have made the changes accordingly and will follow the logic for new changes

**Comment by shashigharti** [comment]
Thank you! @roll . I have made the changes and it is ready for review.

**Review by roll — APPROVED** [review]
Thanks!

File: frictionless/program/validate.py (hunk 2 above)
Comment (shashigharti) [reply frictionless/program/validate.py:286]: I have missed this change. I will fix it in next PR.

**Pull request merged**